src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -1572,13 +1574,26 @@ UndistributeTable(Oid relationId)
 						errdetail("because the table is not distributed")));
 	}
 
-	if (TableReferencing(relationId))
+	bool tableReferencing = TableReferencing(relationId);
+	bool tableReferenced = TableReferenced(relationId);
+	if ((tableReferencing || tableReferenced) && cascade)
+	{
+		ExecuteCitusTableFunctionCascade(relationId, lockMode, UNDISTRIBUTE_TABLE);
```

Comment:
it is not very obvious how partitions with foreign keys would work here. Also, we can probably simplify the code/flow a bit. But in general, it looks possible to stick to this approach

Reply:
>it is not very obvious how partitions with foreign keys would work here.

actually we have a seperate comment for that, do you think that should it be more clear ?

```c
/*
 * Even if we called UndistributeTable with cascade option, we
 * shouldn't cascade via foreign keys on partitions. Otherwise,
 * we might try to undistribute partitions of other tables in
 * our foreign key subgraph more than once.
*/
```

>Also, we can probably simplify the code/flow a bit

I wanted to make code change as small as possible when introducing cascade option. So I put call to `ExecuteCitusTableFunctionCascade` right after foreign key checks.
Do you have any other suggestions to simplify this more ?